test/stress/s3benchmark.cc

```diff
@@ -0,0 +1,225 @@
+  double duration_upload = files_uploaded.tv_sec - payload_generated.tv_sec + 0.000000001*(files_uploaded.tv_nsec - payload_generated.tv_nsec);
+  double duration_reupload = files_reuploaded.tv_sec - files_uploaded.tv_sec + 0.000000001*(files_reuploaded.tv_nsec - files_uploaded.tv_nsec);
+  double duration_delete = files_deleted.tv_sec - files_reuploaded.tv_sec + 0.000000001*(files_deleted.tv_nsec - files_reuploaded.tv_nsec);
+
+  Cleanup();
+
+  std::cout << num_uploads_ << " uploads, " << unique_files_ << " unique files of average size "
+       << avg_file_size_ << " bytes in:" << std::endl
+       << "Payload generation: " << duration_payload << " seconds" << std::endl
+       << "HEAD(NotFound)+PUT(New): " << duration_upload << " seconds" << std::endl
+       << "HEAD(Found)+PUT(Existing): " << duration_reupload << " seconds" << std::endl
+       << "DELETE: " << duration_delete << " seconds" << std::endl;
+}
+
+void S3TestScenario::GeneratePayload() {
+  for(int i = 0; i < num_uploads_; ++i) {
+    float duplicate_chance = rand() * (1.0 / RAND_MAX);
+    if (duplicate_chance < duplicate_file_ratio_ and filenames_.size() > 0) {
+      int duplicate_index = rand() % filenames_.size();
+      filenames_.push_back(filenames_[duplicate_index]);
+    } else {
+      ++unique_files_;
+      filenames_.push_back(GenerateFile());
+    }
+  }
+}
+
+std::string S3TestScenario::GenerateFile()
+{
+  std::string filename;
+  do {
+    filename = GenerateRandomFilename();
+  } while(FileExists(tmp_path_ + "/" + filename));
+  std::string file_path = tmp_path_ + "/" + filename;
+
+  int file_size = 0;
+  while(file_size <= 0) file_size = avg_file_size_ + int(RandomNormal()*avg_file_size_*relative_variance_);
+
+  int chunk_size = 1024;
+  int bytes_written = 0;
+  char *buffer = reinterpret_cast<char *>(smalloc(sizeof(char)*(chunk_size + 1)));
+  buffer[chunk_size] = 0;
+  std::ofstream random_file;
+
+  random_file.open(file_path);
+  assert(!random_file.fail());
+
+  while(bytes_written < file_size) {
+    for(int i = 0; i < chunk_size; ++i) {
+      buffer[i] = rand() % (CHAR_MAX - CHAR_MIN) + CHAR_MIN;
+    }
+    int num_bytes_to_write = (chunk_size <= file_size - bytes_written) ? chunk_size : (file_size - bytes_written);
+    random_file.write(buffer, num_bytes_to_write);
+    bytes_written += num_bytes_to_write;
+  }
+  free(buffer);
+  random_file.close();
+  return filename;
+}
+
+std::string S3TestScenario::GenerateRandomFilename()
+{
+  int len = 10;
+  static const char alphanum[] =
+        "0123456789"
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        "abcdefghijklmnopqrstuvwxyz";
+
+  std::string result(len, 0);
+  for(int i = 0; i < len; ++i) {
+    result[i] = alphanum[rand() % (sizeof(alphanum) - 1)];
+  }
+  return result;
+}
+
+void S3TestScenario::UploadFile(const std::string &filename)
+{
+  std::string file_path = tmp_path_ + "/" + filename;
+  uploader_->Upload(file_path, filename);
```

Comment:
This should rather use the streamed interface of the uploader, which is used for the majority of files.  The `Upload()` routine is only used for the few well-known top-level files like `.cvmfspublished`